Convert set and photo names to valid filenames

astanin opened the pull request:
To avoid exceptions when creating directories and saving files,
only valid filenames should be used.
For portability reasons Windows filename restrictions are enforced:
https://stackoverflow.com/a/31976060/25450

Reserved characters are replaced with an underscore '_'.
Trailing whitespace and dots are removed.
Reserved filenames (like COM or NUL) are prefixed with an underscore.

beaufour commented:
Thanks for the PR! I have a few code comments which I will add in a bit. But is there really not built in way to do this in Python? Ugh. As much as I hate operating system specific code, let's at least only do it on Windows. There is no reason to change the filename on other operating (/file) systems with less restrictions.

beaufour commented on flickr_download/flick_download.py:
```diff
@@ -100,6 +100,23 @@ def download_set(set_id, get_filename, size_label=None, skip_download=False):
+    Unix filesystems allow almost anything but '/' in the filename,
+    but Windows has many restrictions: https://stackoverflow.com/a/31976060/25450
+    """
+    import re
```
please push this to the top of the file

beaufour commented on flickr_download/flick_download.py:
```diff
@@ -100,6 +100,23 @@ def download_set(set_id, get_filename, size_label=None, skip_download=False):
+    but Windows has many restrictions: https://stackoverflow.com/a/31976060/25450
+    """
+    import re
+    forbiddenchars = re.compile(u"([/\\\\<\\>:\"\\|\\?\\*]|[\x00-\x1F])", re.UNICODE)
```
There is no need to `compile` this first, and then do the `sub` since you are only using the expression once. Just do it in one go. That makes it easier to read

beaufour commented on flickr_download/flick_download.py:
```diff
@@ -100,6 +100,23 @@ def download_set(set_id, get_filename, size_label=None, skip_download=False):
     download_list(pset, pset.title, get_filename, size_label, skip_download)
 
 
+def valid_filesystem_name(orig_basename):
```
add `get_` to the function name to follow the style of the rest of the file

astanin commented:
I implemented the changes, mostly as you asked. I moved regular expressions to the global scope. So the same compiled expressions can be reused most of the time.

> But is there really not built in way to do this in Python?

Not that I know of.

> There is no reason to change the filename on other operating (/file) systems with less restrictions.

For a long time I was doing all my photo management in Linux, and eventually I found that it's better to follow the most restrictive file naming conventions, because sooner or later there might be a need to copy files to a Windows filesystem. Basically, just avoiding `<`, `>`, `"`, `:`, slashes, `*`, `?`, and `|` in the filenames makes any archive much more portable.

If you want I may add platform detection checks, but if I were you I would keep the script as simple as possible and choose only one behavior for all platforms.

>  you might also want to treat filenames and directories differently

According to Microsoft docs, naming rules are the same for filenames and directories.

beaufour commented:
Sorry, this got lost.

> I implemented the changes, mostly as you asked. I moved regular expressions to the global scope. So the same compiled expressions can be reused most of the time.
> 
> > But is there really not built in way to do this in Python?
> 
> Not that I know of.

Kind of crazy.

> > There is no reason to change the filename on other operating (/file) systems with less restrictions.
> 
> For a long time I was doing all my photo management in Linux, and eventually I found that it's better to follow the most restrictive file naming conventions, because sooner or later there might be a need to copy files to a Windows filesystem. Basically, just avoiding `<`, `>`, `"`, `:`, slashes, `*`, `?`, and `|` in the filenames makes any archive much more portable.
> 
> If you want I may add platform detection checks, but if I were you I would keep the script as simple as possible and choose only one behavior for all platforms.

I am in favor of platform detection. No reason to limit users on all operating systems because of one platform's restrictions.

> 
> > you might also want to treat filenames and directories differently
> 
> According to Microsoft docs, naming rules are the same for filenames and directories.

Ok. As long as it's platform specific, then all good.

beaufour commented:
I added https://pypi.org/project/pathvalidate/ a while back
